pkg/autoscaler/aggregation/bucketing2.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aggregation
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+)
+
+// TimedFloat64Buckets2 keeps buckets that have been collected at a certain time.
+type TimedFloat64Buckets2 struct {
+	bucketsMutex sync.RWMutex
+	buckets      []float64
+	lastWrite    time.Time
+
+	granularity time.Duration
+	window      time.Duration
+}
+
+func (t *TimedFloat64Buckets2) String() string {
+	return spew.Sdump(t.buckets)
+}
+
+// NewTimedFloat64Buckets2 generates a new TimedFloat64Buckets2 with the given
+// granularity.
+func NewTimedFloat64Buckets2(window, granularity time.Duration) *TimedFloat64Buckets2 {
+	// Number of buckets is `window` divided by `granularity`, rounded up.
+	// e.g. 60s / 2s = 30.
+	nb := int(math.Ceil(float64(window) / float64(granularity)))
+	return &TimedFloat64Buckets2{
+		buckets:     make([]float64, nb),
+		granularity: granularity,
+		window:      window,
+	}
+}
+
+// timeToIndex converts time to an integer that can be used for modulo
+// operations to find the index in the bucket list.
+// bucketMutex needs to be held.
+func (t *TimedFloat64Buckets2) timeToIndex(tm time.Time) int {
+	// I don't think this run in 2038 :-)
+	return int(tm.Unix()) / int(t.granularity.Seconds())
+}
+
+func (t *TimedFloat64Buckets2) reset() {
+	for i := range t.buckets {
+		t.buckets[i] = 0
+	}
+}
+
+// Record adds a value with an associated time to the correct bucket.
+func (t *TimedFloat64Buckets2) Record(now time.Time, name string, value float64) {
+	bucketTime := now.Truncate(t.granularity)
+
+	t.bucketsMutex.Lock()
+	defer t.bucketsMutex.Unlock()
+
+	// NB: we need to divide by granularity, since it's a compressing mapping
+	// to buckets.
+	bucketIdx := t.timeToIndex(now) % len(t.buckets)
```

Comment:
```suggestion
	tIdx := t.timeToIndex(now) % len(t.buckets)
```

for consistency with below's code. Or make them all `bIdx` maybe?